// @phantomit/table/lib/table/FilterDropDown.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __spreadArrays = (this && this.__spreadArrays) || function () {
    for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
    for (var r = Array(s), k = 0, i = 0; i < il; i++)
        for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
            r[k] = a[j];
    return r;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var react_1 = __importDefault(require("react"));
var definition_1 = require("@phantomit/datasource/lib/utils/definition");
var moment_1 = __importDefault(require("moment"));
var antd_1 = require("antd");
exports.NumericFilter = function (props) {
    var _a, _b, _c;
    var _d = react_1.default.useState(1), type = _d[0], setType = _d[1];
    return react_1.default.createElement(react_1.default.Fragment, null,
        react_1.default.createElement(antd_1.Row, { style: { marginBottom: 12 } },
            react_1.default.createElement(antd_1.Radio.Group, { onChange: function (e) { return setType(e.target.value); }, value: type },
                react_1.default.createElement(antd_1.Radio, { value: 1 }, "Exact"),
                react_1.default.createElement(antd_1.Radio, { value: 2 }, "Range"))),
        react_1.default.createElement(antd_1.Row, null,
            type === 1 && react_1.default.createElement(antd_1.InputNumber, { value: (_a = props.values) === null || _a === void 0 ? void 0 : _a.eq, onChange: function (value) {
                    props.onChange({ eq: value });
                } }),
            type === 2 && react_1.default.createElement(react_1.default.Fragment, null,
                react_1.default.createElement(antd_1.Space, null,
                    react_1.default.createElement(antd_1.InputNumber, { value: (_b = props.values) === null || _b === void 0 ? void 0 : _b.lt, placeholder: "less than", onChange: function (value) {
                            var _a;
                            props.onChange({ gt: (_a = props.values) === null || _a === void 0 ? void 0 : _a.gt, lt: value });
                        } }),
                    react_1.default.createElement(antd_1.InputNumber, { value: (_c = props.values) === null || _c === void 0 ? void 0 : _c.gt, placeholder: "greater than", onChange: function (value) {
                            var _a;
                            props.onChange({ lt: (_a = props.values) === null || _a === void 0 ? void 0 : _a.lt, gt: value });
                        } })))));
};
exports.TextFilter = function (props) {
    var _a, _b;
    var _c = react_1.default.useState({ loading: true, data: {} }), opt = _c[0], setOpt = _c[1];
    var load = function () { return __awaiter(void 0, void 0, void 0, function () {
        var data;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, props.distinct()];
                case 1:
                    data = _a.sent();
                    setOpt(__assign(__assign({}, opt), { loading: false, data: data }));
                    return [2 /*return*/];
            }
        });
    }); };
    react_1.default.useEffect(function () {
        if (opt.loading)
            load();
    });
    var items = Object.entries(opt.data).map(function (_a) {
        var key = _a[0], value = _a[1];
        var _b;
        var eq = (_b = props.values) === null || _b === void 0 ? void 0 : _b.eq;
        var checked = eq != null && Array.isArray(eq) && eq.some(function (p) { return key === p; });
        return react_1.default.createElement(antd_1.Row, { key: key },
            react_1.default.createElement(antd_1.Tooltip, { title: value },
                react_1.default.createElement(antd_1.Checkbox, { checked: checked, onChange: function (e) {
                        var _a;
                        var values = props.values != null ? props.values : { eq: [] };
                        var others = values.eq.filter(function (q) { return q !== key; });
                        if (e.target.checked) {
                            props.onChange(__assign(__assign({}, props.values), { eq: (_a = others) === null || _a === void 0 ? void 0 : _a.concat([key]) }));
                        }
                        else {
                            props.onChange(__assign(__assign({}, props.values), { eq: others }));
                        }
                    } }, value)));
    });
    return react_1.default.createElement(react_1.default.Fragment, null,
        react_1.default.createElement(antd_1.Checkbox, { indeterminate: ((_a = props.values) === null || _a === void 0 ? void 0 : _a.eq) != null && props.values.eq.length > 0 && Object.keys(opt.data).length !== props.values.eq.length, checked: ((_b = props.values) === null || _b === void 0 ? void 0 : _b.eq) != null && Object.keys(opt.data).length === props.values.eq.length, onChange: function (e) {
                if (e.target.checked) {
                    props.onChange(__assign(__assign({}, props.values), { eq: Object.keys(opt.data) }));
                }
                else {
                    props.onChange(__assign(__assign({}, props.values), { eq: [] }));
                }
            } }, "Select All"),
        react_1.default.createElement(antd_1.Divider, { style: { margin: 5 } }),
        react_1.default.createElement("div", { style: { maxHeight: 200, overflowY: "auto", overflowX: 'hidden' } }, !opt.loading && items));
};
exports.SelectFilter = function (props) {
    return react_1.default.createElement(react_1.default.Fragment, null);
};
exports.DateFilter = function (props) {
    var _a, _b, _c;
    var _d = react_1.default.useState(1), type = _d[0], setType = _d[1];
    var formatter = function (date) {
        var out = props.dateOut;
        if (!date)
            return "";
        if (out) {
            if (typeof out === "function") {
                return out(date);
            }
            else {
                return moment_1.default(date).format(out);
            }
        }
        return date;
    };
    return react_1.default.createElement(react_1.default.Fragment, null,
        react_1.default.createElement(antd_1.Row, { style: { marginBottom: 12 } },
            react_1.default.createElement(antd_1.Radio.Group, { onChange: function (e) { return setType(e.target.value); }, value: type },
                react_1.default.createElement(antd_1.Radio, { value: 1 }, "Exact"),
                react_1.default.createElement(antd_1.Radio, { value: 2 }, "Range"))),
        react_1.default.createElement(antd_1.Row, null,
            type === 1 &&
                react_1.default.createElement(antd_1.DatePicker, { value: ((_a = props.values) === null || _a === void 0 ? void 0 : _a.eq) == null ? null : moment_1.default(props.values.eq), onChange: function (value) {
                        var _a;
                        props.onChange({ eq: formatter((_a = value) === null || _a === void 0 ? void 0 : _a.format()) });
                    } }),
            type === 2 && react_1.default.createElement(antd_1.DatePicker.RangePicker, { defaultValue: [moment_1.default((_b = props.values) === null || _b === void 0 ? void 0 : _b.lt), moment_1.default((_c = props.values) === null || _c === void 0 ? void 0 : _c.gt)], onChange: function (dates, str) {
                    props.onChange({ lt: formatter(str[0]), gt: formatter(str[1]) });
                } })));
};
exports.BooleanFilter = function (props) {
    var _a = (props.values ? props.values : {}).eq, eq = _a === void 0 ? [] : _a;
    return react_1.default.createElement(react_1.default.Fragment, null,
        react_1.default.createElement(antd_1.Row, null,
            react_1.default.createElement(antd_1.Checkbox, { indeterminate: eq.length === 1, checked: eq.length === 2, onChange: function (e) {
                    if (e.target.checked)
                        props.onChange({ eq: [false, true] });
                    else
                        props.onChange({ eq: [] });
                } }, "All")),
        react_1.default.createElement(antd_1.Divider, { style: { margin: 5 } }),
        react_1.default.createElement(antd_1.Row, null,
            react_1.default.createElement(antd_1.Checkbox, { checked: eq.includes(true), onChange: function (e) {
                    if (e.target.checked)
                        props.onChange({ eq: __spreadArrays(eq, [true]) });
                    else
                        props.onChange({ eq: eq.filter(function (p) { return p !== true; }) });
                } }, "True")),
        react_1.default.createElement(antd_1.Row, null,
            react_1.default.createElement(antd_1.Checkbox, { checked: eq.includes(false), onChange: function (e) {
                    if (e.target.checked)
                        props.onChange({ eq: __spreadArrays(eq, [false]) });
                    else
                        props.onChange({ eq: eq.filter(function (p) { return p !== false; }) });
                } }, "False")));
};
exports.FilterDropDown = function (props) {
    var _a = react_1.default.useState(props.selected), filter = _a[0], setFilter = _a[1];
    var type = filter && ((filter.eq && Array.isArray(filter.eq)) || filter.lt || filter.gt) ? "multiple" : "single";
    return react_1.default.createElement(react_1.default.Fragment, null,
        react_1.default.createElement("div", { style: { padding: 12 } },
            props.column.type === definition_1.ColumnType.boolean && react_1.default.createElement(exports.BooleanFilter, { values: filter, type: type, onChange: setFilter, distinct: props.distinct }),
            props.column.type === definition_1.ColumnType.int && react_1.default.createElement(exports.NumericFilter, { values: filter, type: type, onChange: setFilter, distinct: props.distinct }),
            props.column.type === definition_1.ColumnType.float && react_1.default.createElement(exports.NumericFilter, { values: filter, type: type, onChange: setFilter, distinct: props.distinct }),
            props.column.type === definition_1.ColumnType.currency && react_1.default.createElement(exports.NumericFilter, { values: filter, type: type, onChange: setFilter, distinct: props.distinct }),
            props.column.type === definition_1.ColumnType.date && react_1.default.createElement(exports.DateFilter, { dateOut: props.column.dateOut, values: filter, type: type, onChange: setFilter, distinct: props.distinct }),
            props.column.type === definition_1.ColumnType.string && react_1.default.createElement(exports.TextFilter, { values: filter, type: type, onChange: function (f) {
                    setFilter(f);
                }, distinct: props.distinct }),
            react_1.default.createElement(antd_1.Divider, { style: { margin: 8 } }),
            react_1.default.createElement(antd_1.Space, { align: "end" },
                react_1.default.createElement(antd_1.Button, { onClick: function () {
                        props.clear();
                        setFilter({});
                    } }, "Clear"),
                react_1.default.createElement(antd_1.Button, { type: "primary", onClick: function () {
                        props.filter(filter);
                    } }, "Apply"))));
};
