// @phantomit/table/lib/table/Table.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var react_1 = __importDefault(require("react"));
var TableWrap_1 = require("./TableWrap");
var Datasource_1 = require("@phantomit/datasource/lib/Datasource");
exports.Table = function (props) {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l;
    var _m = react_1.default.useState({ init: true, columns: [] }), opt = _m[0], setOpt = _m[1];
    var load = react_1.default.useCallback(function () { return __awaiter(void 0, void 0, void 0, function () {
        var columns;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, props.columns()];
                case 1:
                    columns = _a.sent();
                    setOpt(__assign(__assign({}, opt), { init: false, columns: columns }));
                    return [2 /*return*/];
            }
        });
    }); }, [opt, props]);
    react_1.default.useEffect(function () {
        if (opt.init) {
            console.log("init");
            load();
        }
    }, [load, opt.init]);
    return react_1.default.createElement(react_1.default.Fragment, null, !opt.init && react_1.default.createElement(TableWrap_1.TableWrap, { source: props.source, server: {
            fetch: function () { return __awaiter(void 0, void 0, void 0, function () {
                var data;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0: return [4 /*yield*/, props.fetch()];
                        case 1:
                            data = _a.sent();
                            console.log(data);
                            return [4 /*yield*/, Datasource_1.toRow(data)];
                        case 2: return [2 /*return*/, _a.sent()];
                    }
                });
            }); },
            save: function (row) { return __awaiter(void 0, void 0, void 0, function () {
                var __id__, args, data;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            __id__ = row.__id__, args = __rest(row, ["__id__"]);
                            console.log(row);
                            return [4 /*yield*/, props.save(__assign({}, args))];
                        case 1:
                            data = _a.sent();
                            return [2 /*return*/, __assign(__assign({}, data), { __id__: __id__ })];
                    }
                });
            }); },
            delete: function (row) { return __awaiter(void 0, void 0, void 0, function () {
                var __id__, args, data;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            __id__ = row.__id__, args = __rest(row, ["__id__"]);
                            return [4 /*yield*/, props.delete(__assign({}, args))];
                        case 1:
                            data = _a.sent();
                            return [2 /*return*/, __assign(__assign({}, data), { __id__: __id__ })];
                    }
                });
            }); }
        }, cfg: {
            onColumnUpdate: props.saveColumn,
            minWidth: props.minWidth,
            actionColumnName: (_a = props.actions) === null || _a === void 0 ? void 0 : _a.columnName,
            actionColumnWidth: (_b = props.actions) === null || _b === void 0 ? void 0 : _b.width,
            columns: opt.columns,
            tableName: props.title,
            select: (_c = props.actions) === null || _c === void 0 ? void 0 : _c.select,
            refresh: (_d = props.actions) === null || _d === void 0 ? void 0 : _d.refresh,
            add: (_e = props.actions) === null || _e === void 0 ? void 0 : _e.add,
            delete: (_f = props.actions) === null || _f === void 0 ? void 0 : _f.delete,
            edit: (_g = props.actions) === null || _g === void 0 ? void 0 : _g.edit,
            settings: (_h = props.actions) === null || _h === void 0 ? void 0 : _h.settings,
            export: (_j = props.actions) === null || _j === void 0 ? void 0 : _j.export,
            import: (_k = props.actions) === null || _k === void 0 ? void 0 : _k.import,
            search: (_l = props.actions) === null || _l === void 0 ? void 0 : _l.search,
            footer: props.footer,
            header: props.header,
            buttons: props.headerButtons,
            customActions: props.customActions,
        }, rowClass: props.rowClass, validateAction: props.validateAction, scrollX: props.scrollX, scrollY: props.scrollY, loading: props.loading }));
};
