// @phantomit/datasource/lib/v2/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var DataSourceImpl_1 = require("./DataSourceImpl");
var Sort_1 = require("./Sort");
var Filters_1 = require("./Filters");
var Formatter_1 = require("./Formatter");
exports.default = {
    Datasource: DataSourceImpl_1.DataSourceImpl,
    FilterSet: DataSourceImpl_1.FilterSet,
    PropertyId: DataSourceImpl_1.PropertyId,
    Sort: {
        SortBy: Sort_1.SortBy, Then: Sort_1.Then
    },
    Filters: {
        And: Filters_1.And, Any: Filters_1.Any, Contains: Filters_1.Contains, NotContains: Filters_1.NotContains, Equal: Filters_1.Equal, GreaterThanOrEqual: Filters_1.GreaterThanOrEqual, GreaterThan: Filters_1.GreaterThan, LessThanOrEqual: Filters_1.LessThanOrEqual, LessThan: Filters_1.LessThan, Not: Filters_1.Not, Or: Filters_1.Or
    },
    Format: {
        Combine: Formatter_1.Combine, ArrayFormat: Formatter_1.ArrayFormat, SuffixFormat: Formatter_1.SuffixFormat, PrefixFormat: Formatter_1.PrefixFormat, DateFormat: Formatter_1.DateFormat, NumberFormat: Formatter_1.NumberFormat, CurrencyFormat: Formatter_1.CurrencyFormat, EnumFormat: Formatter_1.EnumFormat, createFormatter: Formatter_1.createFormatter
    }
};
